File: packages/sdk-go/dedup_test.go
package rollgate

import (
	"errors"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

func TestRequestDeduplicator_Dedupe(t *testing.T) {
	t.Run("should execute function and return result", func(t *testing.T) {
		dedup := NewRequestDeduplicator()
		callCount := 0

		result, err := dedup.Dedupe("key", func() (any, error) {
			callCount++
			return "result", nil
		})

		if err != nil {
			t.Errorf("expected no error, got %v", err)
		}
		if result != "result" {
			t.Errorf("expected 'result', got %v", result)
		}
		if callCount != 1 {
			t.Errorf("expected function called once, got %d", callCount)
		}
	})

	t.Run("should deduplicate concurrent requests", func(t *testing.T) {
		dedup := NewRequestDeduplicator()
		var callCount int32

		var wg sync.WaitGroup
		results := make([]any, 3)
		errors := make([]error, 3)

		for i := 0; i < 3; i++ {
			wg.Add(1)
			go func(idx int) {
				defer wg.Done()
				results[idx], errors[idx] = dedup.Dedupe("key", func() (any, error) {
					atomic.AddInt32(&callCount, 1)
					time.Sleep(50 * time.Millisecond)
					return "result", nil
				})
			}(i)
		}

		wg.Wait()

		for i, err := range errors {
			if err != nil {
				t.Errorf("request %d got error: %v", i, err)
			}
		}

		for i, result := range results {
			if result != "result" {
				t.Errorf("request %d expected 'result', got %v", i, result)
			}
		}

		if callCount != 1 {
			t.Errorf("expected function called once, got %d", callCount)
		}
	})

	t.Run("should allow new requests after previous completes", func(t *testing.T) {
		dedup := NewRequestDeduplicator()
		callCount := 0

		_, _ = dedup.Dedupe("key", func() (any, error) {
			callCount++
			return "result1", nil
		})

		_, _ = dedup.Dedupe("key", func() (any, error) {
			callCount++
			return "result2", nil
		})

		if callCount != 2 {
			t.Errorf("expected function called twice, got %d", callCount)
		}
	})

	t.Run("should handle different keys independently", func(t *testing.T) {
		dedup := NewRequestDeduplicator()
		var callCount1, callCount2 int32

		var wg sync.WaitGroup
		wg.Add(2)

		go func() {
			defer wg.Done()
			_, _ = dedup.Dedupe("key1", func() (any, error) {
				atomic.AddInt32(&callCount1, 1)
				time.Sleep(10 * time.Millisecond)
				return "result1", nil
			})
		}()

		go func() {
			defer wg.Done()
			_, _ = dedup.Dedupe("key2", func() (any, error) {
				atomic.AddInt32(&callCount2, 1)
				time.Sleep(10 * time.Millisecond)
				return "result2", nil
			})
		}()

		wg.Wait()

		if callCount1 != 1 {
			t.Errorf("expected key1 called once, got %d", callCount1)
		}
		if callCount2 != 1 {
			t.Errorf("expected key2 called once, got %d", callCount2)
		}
	})

	t.Run("should propagate errors to all callers", func(t *testing.T) {
		dedup := NewRequestDeduplicator()
		expectedErr := errors.New("test error")
		var callCount int32

		var wg sync.WaitGroup
		errs := make([]error, 3)

		for i := 0; i < 3; i++ {
			wg.Add(1)
			go func(idx int) {
				defer wg.Done()
				_, errs[idx] = dedup.Dedupe("key", func() (any, error) {
					atomic.AddInt32(&callCount, 1)
					time.Sleep(50 * time.Millisecond)
					return nil, expectedErr
				})
			}(i)
		}

		wg.Wait()

		for i, err := range errs {
			if err != expectedErr {
				t.Errorf("request %d expected error %v, got %v", i, expectedErr, err)
			}
		}

		if callCount != 1 {
			t.Errorf("expected function called once, got %d", callCount)
		}
	})
}

func TestRequestDeduplicator_IsInflight(t *testing.T) {
	t.Run("should return true while request is in-flight", func(t *testing.T) {
		dedup := NewRequestDeduplicator()
		started := make(chan struct{})
		done := make(chan struct{})

		go func() {
			_, _ = dedup.Dedupe("key", func() (any, error) {
				close(started)
				<-done
				return "result", nil
			})
		}()

		<-started

		if !dedup.IsInflight("key") {
			t.Error("expected IsInflight to be true")
		}

		close(done)
		time.Sleep(10 * time.Millisecond)

		if dedup.IsInflight("key") {
			t.Error("expected IsInflight to be false after completion")
		}
	})

	t.Run("should return false for unknown key", func(t *testing.T) {
		dedup := NewRequestDeduplicator()

		if dedup.IsInflight("unknown") {
			t.Error("expected IsInflight to be false for unknown key")
		}
	})
}

func TestRequestDeduplicator_Clear(t *testing.T) {
	dedup := NewRequestDeduplicator()
	started := make(chan struct{})

	go func() {
		_, _ = dedup.Dedupe("key", func() (any, error) {
			close(started)
			time.Sleep(100 * time.Millisecond)
			return "result", nil
		})
	}()

	<-started

	if !dedup.IsInflight("key") {
		t.Error("expected IsInflight to be true before clear")
	}

	dedup.Clear()

	if dedup.IsInflight("key") {
		t.Error("expected IsInflight to be false after clear")
	}
}

File: packages/sdk-core/src/dedup.test.ts
import { describe, it, expect, vi } from 'vitest';
import { RequestDeduplicator } from './dedup';

describe('RequestDeduplicator', () => {
  describe('dedupe', () => {
    it('should execute function and return result', async () => {
      const dedup = new RequestDeduplicator();
      const fn = vi.fn().mockResolvedValue('result');

      const result = await dedup.dedupe('key', fn);

      expect(result).toBe('result');
      expect(fn).toHaveBeenCalledTimes(1);
    });

    it('should deduplicate concurrent identical requests', async () => {
      const dedup = new RequestDeduplicator();
      const fn = vi
        .fn()
        .mockImplementation(
          () => new Promise((resolve) => setTimeout(() => resolve('result'), 100))
        );

      const promise1 = dedup.dedupe('key', fn);
      const promise2 = dedup.dedupe('key', fn);
      const promise3 = dedup.dedupe('key', fn);

      const results = await Promise.all([promise1, promise2, promise3]);

      expect(results).toEqual(['result', 'result', 'result']);
      expect(fn).toHaveBeenCalledTimes(1);
    });

    it('should allow new requests after previous completes', async () => {
      const dedup = new RequestDeduplicator();
      const fn = vi.fn().mockResolvedValue('result');

      await dedup.dedupe('key', fn);
      await dedup.dedupe('key', fn);

      expect(fn).toHaveBeenCalledTimes(2);
    });

    it('should handle different keys independently', async () => {
      const dedup = new RequestDeduplicator();
      const fn1 = vi.fn().mockResolvedValue('result1');
      const fn2 = vi.fn().mockResolvedValue('result2');

      const results = await Promise.all([dedup.dedupe('key1', fn1), dedup.dedupe('key2', fn2)]);

      expect(results).toEqual(['result1', 'result2']);
      expect(fn1).toHaveBeenCalledTimes(1);
      expect(fn2).toHaveBeenCalledTimes(1);
    });

    it('should propagate errors to all callers', async () => {
      const dedup = new RequestDeduplicator();
      const error = new Error('test error');
      const fn = vi.fn().mockRejectedValue(error);

      const promise1 = dedup.dedupe('key', fn);
      const promise2 = dedup.dedupe('key', fn);

      await expect(promise1).rejects.toThrow('test error');
      await expect(promise2).rejects.toThrow('test error');
      expect(fn).toHaveBeenCalledTimes(1);
    });

    it('should clean up after error', async () => {
      const dedup = new RequestDeduplicator();
      const fn = vi.fn().mockRejectedValueOnce(new Error('error')).mockResolvedValue('success');

      await dedup.dedupe('key', fn).catch(() => {});
      const result = await dedup.dedupe('key', fn);

      expect(result).toBe('success');
      expect(fn).toHaveBeenCalledTimes(2);
    });
  });

  describe('isInflight', () => {
    it('should return true while request is in-flight', async () => {
      const dedup = new RequestDeduplicator();
      let resolve: (value: string) => void;
      const fn = vi.fn().mockImplementation(
        () =>
          new Promise((r) => {
            resolve = r;
          })
      );

      const promise = dedup.dedupe('key', fn);

      expect(dedup.isInflight('key')).toBe(true);

      resolve!('result');
      await promise;

      expect(dedup.isInflight('key')).toBe(false);
    });

    it('should return false for unknown key', () => {
      const dedup = new RequestDeduplicator();
      expect(dedup.isInflight('unknown')).toBe(false);
    });
  });

  describe('clear', () => {
    it('should clear all in-flight tracking', () => {
      const dedup = new RequestDeduplicator();
      const fn = vi.fn().mockImplementation(
        () => new Promise(() => {}) // Never resolves
      );

      dedup.dedupe('key1', fn);
      dedup.dedupe('key2', fn);

      expect(dedup.isInflight('key1')).toBe(true);
      expect(dedup.isInflight('key2')).toBe(true);

      dedup.clear();

      expect(dedup.isInflight('key1')).toBe(false);
      expect(dedup.isInflight('key2')).toBe(false);
    });
  });
});
